challenge_8/C/randomDeepCopy.c

```diff
@@ -0,0 +1,111 @@
+/*
+Author: karanchawla
+Random Deep Copy
+#Challenge #8
+*/
+
+#include <stdio.h> 
+#include <stdlib.h> 
+
+
+//node definition for the linked list
+struct node
+{
+	int data;
+	struct node* next;
+	struct node* random;
+};
+
+typedef struct node* Node;
+
+//utility function to create a node
+Node newNode(int key)
+{
+	Node temp = (struct node*) malloc(sizeof(struct node));
+
+	temp->data = key;
+	temp->next = NULL;
+	temp->random = NULL;
+
+	return temp;
+}
+
+//utility function to print the list
+void printList(Node head)
+{
+	while(head!=NULL)
+	{
+		printf("Node key: %d\t", head->data);
+		printf("Random Node pointer data: %d\n", head->random->data);
+		head = head->next;
+	}
+	return;
+}
+
+//First step is to insert a copy of each node
+//after the first node in the original list
+//Copy the random pointers
+//Separate this second list
+Node deepCopy(Node head)
+{
+	Node old;
+	Node temp = head;
+	//modify this list
+	while(temp)
+	{
+
+		Node temp2 = newNode(temp->data);
+		old = temp->next;
+		temp->next = temp2;
+		temp2->next = old;
+		temp = old;
+	}
+
+	temp = head;
+	//copy arbitrary matching
+	while(temp && temp->next && temp->random)
+	{
+
+		temp->next->random = temp->random->next;
+		temp = temp->next->next;
+	}
+
+	//Separating the two lists
+	Node newNode = head->next;
+	temp = head; 
+
+	while(temp)
+	{
+		old = temp->next;
+		temp->next = old->next;
+		temp = old->next;
+		if(old->next) old->next->next;
+	}
+
+	return newNode;
+}
```

Comment:
You also mix immediate and referenced objects in here. `Node tmp` is accessed with `tmp.field`; `Node* other` is accessed with `other->field`.

`Node temp` is absolutely correct, but you may want to double check what should be references (`Node*`) vs immediates (`Node`) in this.